### ISSUE TITLE
inference result very bad although train was successful

### ISSUE THREAD

**Issue by aditya-jindal-trysolvio** [BADBADBADBOY/pytorchOCR#24]
I trained a model and got 82% accuracy on train set. I then used a train set image for inference and the result was very bad. 
Here are the logs:
`loss_total	loss_score	loss_border	loss_tvo	loss_tco	piexl_acc	piexl_iou	t_rescall	t_precision	t_hmean	b_rescall	b_precision	b_hmean	
13.674375	0.570150	0.150688	6.066021	2.569671	0.762514	0.496774	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	
8.445973	0.510070	0.128200	4.033121	1.172013	0.805130	0.536731	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	
5.466000	0.485222	0.118125	2.611230	0.630538	0.815070	0.560080	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	
3.890232	0.458884	0.113267	1.757485	0.454570	0.821011	0.585244	0.000000	0.000000	0.000000	0.000000	0.000000	0.000000	
`
I didn't use validation so there are 0 in some of the parameters.
 ![image](https://user-images.githubusercontent.com/79012105/108401194-18896e80-7242-11eb-94f5-aa216acbd1cc.png) 
The result i got which is very wrong. Is there some problem with inference script which is causing this. There was one problem with the script about which i have opened an issue and have commented the fix.

The inference code i used `python3 ./tools/det_infer.py --config ./config/det_SAST_resnet50.yaml --model_path ./checkpoint/ag_SAST_bb_resnet50_he_SASTHead_bs_2_ep_600_None/SAST_3.pth.tar --img_path ./test_images/template_batch1_thread5_2939.jpg  --result_save_path ./result --batch_size 2 --max_size 1024`. I trained the model with img_size (1024,1024).

**Comment by fxwfzsxyq** [comment]
> I trained a model and got 82% accuracy on train set. I then used a train set image for inference and the result was very bad.
> Here are the logs:
> `loss_total loss_score loss_border loss_tvo loss_tco piexl_acc piexl_iou t_rescall t_precision t_hmean b_rescall b_precision b_hmean 13.674375 0.570150 0.150688 6.066021 2.569671 0.762514 0.496774 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 8.445973 0.510070 0.128200 4.033121 1.172013 0.805130 0.536731 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 5.466000 0.485222 0.118125 2.611230 0.630538 0.815070 0.560080 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 3.890232 0.458884 0.113267 1.757485 0.454570 0.821011 0.585244 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 `
> I didn't use validation so there are 0 in some of the parameters.
> ![image](https://user-images.githubusercontent.com/79012105/108401194-18896e80-7242-11eb-94f5-aa216acbd1cc.png)
> The result i got which is very wrong. Is there some problem with inference script which is causing this. There was one problem with the script about which i have opened an issue and have commented the fix.
> 
> The inference code i used `python3 ./tools/det_infer.py --config ./config/det_SAST_resnet50.yaml --model_path ./checkpoint/ag_SAST_bb_resnet50_he_SASTHead_bs_2_ep_600_None/SAST_3.pth.tar --img_path ./test_images/template_batch1_thread5_2939.jpg --result_save_path ./result --batch_size 2 --max_size 1024`. I trained the model with img_size (1024,1024).


if --img_path is a image file ,the batch_size should be 1

**Comment by aditya-jindal-trysolvio** [comment]
still same result after trying your command. 
![template_batch1_thread5_2939](https://user-images.githubusercontent.com/79012105/108473402-fdefdd80-72b3-11eb-9a24-566a8269cb8c.jpg)
Is there some problem with resizing which is not making the bbox correctly fit the image. The image size is 597,894 and the model input is (1024,1024).

**Comment by fxwfzsxyq** [comment]
> still same result after trying your command.
> ![template_batch1_thread5_2939](https://user-images.githubusercontent.com/79012105/108473402-fdefdd80-72b3-11eb-9a24-566a8269cb8c.jpg)
> Is there some problem with resizing which is not making the bbox correctly fit the image. The image size is 597,894 and the model input is (1024,1024).

change the code in line 127-128：

if(self.config['base']['algorithm']=='SAST'):
            scales = [(1.0/scales[i][1],1.0/scales[i][0],ori_imgs[i].shape[0],ori_imgs[i].shape[1]) for i in range(len(scales))]

**Comment by fxwfzsxyq** [comment]
this is my result
![image](https://user-images.githubusercontent.com/18416103/108474547-7feb0100-72ca-11eb-96e8-0ffe3d5ed71d.png)

**Comment by fxwfzsxyq** [comment]
I recommend DBNet for good results and fast speed

**Comment by hailcpy** [comment]
Your result looks very good. I am still getting the same result after doing the change you mentioned. I tried some other images from train set and they give bad results too. The training went to 85% accuracy and loss total was 3.102.. , so i can't understand why so bad results.
![test](https://user-images.githubusercontent.com/60291306/108479430-ba996d00-72bb-11eb-9813-99b552972017.jpg)

**Comment by aditya-jindal-trysolvio** [comment]
Your result looks very good. I am still getting the same result after doing the change you mentioned. I tried some other images from train set and they give bad results too. The training went to 85% accuracy and loss total was 3.102.. , so i can't understand why so bad results.
![test](https://user-images.githubusercontent.com/60291306/108479430-ba996d00-72bb-11eb-9813-99b552972017.jpg)

**Comment by aditya-jindal-trysolvio** [comment]
can u inference with your trained sast model on this:
![template_batch1_thread5_2939 (3)](https://user-images.githubusercontent.com/60291306/108480192-a6a23b00-72bc-11eb-9e3b-239c0a5b8a85.jpg)

**Comment by fxwfzsxyq** [comment]
> Your result looks very good. I am still getting the same result after doing the change you mentioned. I tried some other images from train set and they give bad results too. The training went to 85% accuracy and loss total was 3.102.. , so i can't understand why so bad results.
> ![test](https://user-images.githubusercontent.com/60291306/108479430-ba996d00-72bb-11eb-9813-99b552972017.jpg)


this my loss
![image](https://user-images.githubusercontent.com/18416103/108480297-bd06c180-72d1-11eb-914e-d6f10d4b8975.png)

**Comment by fxwfzsxyq** [comment]
I don’t know if it’s your training problem or something else. If you can’t solve it, I suggest you try to use dbnet first. Many people have tried this without any problems. I will sort out the sast part later. There may be some problems in this part.

**Comment by hailcpy** [comment]
Hey after this line https://github.com/BADBADBADBOY/pytorchOCR/blob/049a4b3f3d9502eb3fc8917c8353edeceb3c9d65/tools/det_infer.py#L97
if i check the img.shape i get torch(3,1024,768). I trained the model on [1024,1024] img_size so shouldn't this be torch(3,1024,1024).

**Comment by aditya-jindal-trysolvio** [comment]
Hey after this line https://github.com/BADBADBADBOY/pytorchOCR/blob/049a4b3f3d9502eb3fc8917c8353edeceb3c9d65/tools/det_infer.py#L97
if i check the img.shape i get torch(3,1024,768). I trained the model on [1024,1024] img_size so shouldn't this be torch(3,1024,1024).

**Comment by hailcpy** [comment]
@fxwfzsxyq  I had done some problem in the dataset. fixed it, now it's working fine. Thanks

**Comment by aditya-jindal-trysolvio** [comment]
@fxwfzsxyq  I had done some problem in the dataset. fixed it, now it's working fine. Thanks for your help.
